File: ios/Classes/NativeTabBarViewFactory.swift
//
//  NativeTabBarViewFactory.swift
//  native_tab_bar_ios
//
//  Created by Grushka, Tom on 9/22/22.
//

import Flutter
import Foundation

class NativeTabBarViewFactory: NSObject, FlutterPlatformViewFactory {
    let hostApi: NativeTabBarHostApiIOS
    let flutterApi: NativeTabBarFlutterApi
    let registrar: FlutterPluginRegistrar

    init(hostApi: NativeTabBarHostApiIOS, flutterApi: NativeTabBarFlutterApi, registrar: FlutterPluginRegistrar) {
        self.hostApi = hostApi
        self.flutterApi = flutterApi
        self.registrar = registrar
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        FlutterJSONMessageCodec()
    }

    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        guard
            let args = args as? [String : Any?],
            let id = args["id"] as? String
        else {
            preconditionFailure("Instance ID must be passed from Flutter via JSONMessageCodec as { 'id': String }")
        }
        return NativeTabBarViewController(frame: frame, id: id, viewId: viewId, hostApi: hostApi, flutterApi: flutterApi, registrar: registrar)
    }
}

File: android/src/main/kotlin/com/dra11y/flutter/native_tab_bar/NativeTabBarViewFactory.kt
package com.dra11y.flutter.native_tab_bar

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.JSONMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import org.json.JSONObject

class NativeTabBarViewFactory(
    private val hostApi: NativeTabBarHostApiAndroid,
    private val flutterApi: NativeTabBarFlutterApi,
    private val flutterAssets: FlutterPlugin.FlutterAssets,
) : PlatformViewFactory(JSONMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val id = (args as? JSONObject)?.optString("id") ?:
            throw Error("Instance ID must be passed from Flutter via JSONMessageCodec as { 'id': String }")
        return NativeTabBarViewController(
            id = id,
            context = context,
            viewId = viewId,
            hostApi = hostApi,
            flutterApi = flutterApi,
            flutterAssets = flutterAssets,
        )
    }
}
